static/app/components/lazyLoad.tsx

```diff
@@ -12,137 +12,111 @@ type PromisedImport<C> = Promise<{default: C}>;
+ */
+function LazyLoad<C extends ComponentType>(props: Props<C>) {
+  const [LazyComponent, setLazyComponent] = useState<C | null>(null);
+  const [error, setError] = useState<any>(null);
+
+  const handleFetchError = useCallback(
+    (fetchError: any) => {
+      Sentry.withScope(scope => {
+        if (isWebpackChunkLoadingError(fetchError)) {
+          scope.setFingerprint(['webpack', 'error loading chunk']);
+        }
+        Sentry.captureException(fetchError);
+      });
+
+      // eslint-disable-next-line no-console
+      console.error(fetchError);
+      setError(fetchError);
+    },
+    [setError]
+  );
+
+  const importComponent = props.component ?? props.route?.componentPromise;
+
+  const fetchComponent = useCallback(async () => {
+    if (importComponent === undefined) {
       return;
     }
 
-    // This is to handle the following case:
-    // <Route path="a/">
-    //   <Route path="b/" component={LazyLoad} componentPromise={...} />
-    //   <Route path="c/" component={LazyLoad} componentPromise={...} />
-    // </Route>
-    //
-    // `LazyLoad` will get not fully remount when we switch between `b` and `c`,
-    // instead will just re-render.  Refetch if route paths are different
-    if (nextProps.route && nextProps.route === this.props.route) {
-      return;
-    }
+    // If we're refetching due to a change to importComponent we want to make
+    // sure the current LazyComponent is cleared out.
+    setLazyComponent(null);
 
-    // If `this.fetchComponent` is not in callback,
-    // then there's no guarantee that new Component will be rendered
-    this.setState(
-      {
-        LazyComponent: null,
-      },
-      this.fetchComponent
+    try {
+      const resolvedComponent = await retryableImport(importComponent);
+
+      // XXX: Because the resolvedComponent may be a functional component (a
+      // function) trying to pass it into the setLazyComponent will cause the
+      // useState to try and execute the function (because useState provides a
+      // "functional updates" invocation, see [0]) which is NOT what we want.
+      // So we use a functional update invocation to set the component.
+      //
+      // [0]: https://reactjs.org/docs/hooks-reference.html#functional-updates
+      setLazyComponent(() => resolvedComponent);
+    } catch (err) {
+      handleFetchError(err);
+    }
+  }, [importComponent, handleFetchError]);
+
+  // Fetch the component on mount and if the importComponent is updated
+  useEffect(() => void fetchComponent(), [fetchComponent]);
+
+  const fetchRetry = useCallback(() => {
+    setError(null);
+    fetchComponent();
+  }, [setError, fetchComponent]);
+
+  if (error) {
+    return (
+      <LoadingErrorContainer>
+        <LoadingError
+          onRetry={fetchRetry}
+          message={t('There was an error loading a component.')}
+        />
+      </LoadingErrorContainer>
     );
   }
 
-  componentDidCatch(error: any) {
```

Comment:
So we lost this. But this component is already typically wrapped in `errorHandler`: https://github.com/getsentry/sentry/blob/b8a692fdc31256b4374c58791d463fdd672e885b/static/app/routes.tsx#L61

if anything, we might be able to just get away with having the default export _always_ be wrapped in this.

There's only one usage outside of routes right now, which is here

https://github.com/getsentry/sentry/blob/b8a692fdc31256b4374c58791d463fdd672e885b/static/app/components/events/rrwebIntegration.tsx#L59-L64

(@billyvg I wonder if we should switch this over to suspense? It's the _only_ thing using LazyLoad that isn't in routes.tsx)

Reply:
Seems fine to lose if only route components ever use it. The benefit of error boundaries is generally for smaller components anyway imo